### DISCUSSION TITLE
How can I detect rotated images/documents?

### DISCUSSION THREAD

**Discussion by denivic** [mindee/doctr#1283] [Q&A] [answered]
Hello,

I'm sort of new to AI. While I have had two courses about machine learning at university, I never really applied the knowledge outside of the class and since it's been 2 years since I finished my BSc, I kind of forgot most of it. I hope you guys can help me.

I would like to detect whether a document has been rotated or not, which I believe is what `db_resnet50_rotation` is for, right? Well, I have had _some_ success using this model for extracting text on a rotated image, but the extracted strings often don't match the text on the document. I'm unsure of how I can improve the accuracy, so what I would like to do instead is detect whether a PDF is rotated and if it is then I would like to change its rotation such that it's "straight" after which I will use EasyOCR, which I've had great success with.

I have tried the following using PDFs that I converted to images:
```py
def get_orientation(self, image):
predictor = detection_predictor('db_resnet50_rotation',
                                pretrained=True,
                                assume_straight_pages=False,
                                preserve_aspect_ratio=True)
result = predictor(image)
return result
```

I was expecting to get a normalized value between 0 and 1, but instead I get a multidimensional array with float values. Initially I thought it returned the image itself so I tried to display it, but on second thought it doesn't make sense to return an image. So what am I missing here? What am I doing wrong?

I tried to look at the documentation, but the documentation wasn't all that helpful. It says `detection_predictor`returns a `predictor`. But how is predictor used and what can I expect it to return? It doesn't say that.

My current solution is to use PyTesseract, but I would prefer to use one module rather than multiple:
```py
def get_orientation(self, path: str) -> int:
    metadata = pytesseract.image_to_osd(path)
    orientation = search('(?<=Rotate: )\d+', metadata).group(0)
    return orientation
```

Ideally I would like to train my own model on the documents we use as they have a specific layout and training on them might improve our results. We have a server available with AI-centric GPUs. This might be more appropriate in another thread though. 

**Side note:** I don't know about you, but the name makes it hard to search for help as Google constantly correct it to "doctor". Even with quotation marks it is still hard to find answers.

**Comment by felixdittrich92** [comment, marked as answer]
Hi @denivic :wave:,

normally from my experience docTR works in the most cases more accure as EasyOCR would it be possible to share an example image to see what's going wrong ? :)

If you want to get the document orientation you can simply pass `detect_orientation=True` to the `ocr_predictor` (with ocr) otherwise you can use the following function: https://github.com/mindee/doctr/blob/3deac68a244952feb6bb6540453d98a9f51dde6f/doctr/models/_utils.py#L29

The `detection_predictor` returns the normalized boxes -> (113, 4, 2)  (num_boxes, (coords - [[x1 y1], [x2 y2], [x3 y3], [x4 y4]])) to restore the values you need to multiply the values with the corresponding page dimensions (width, height).

If you can train your own models see: https://mindee.github.io/doctr/latest/using_doctr/custom_models_training.html

Best regards,
Felix

**Comment by denivic** [comment]
Hey @felixdittrich92!

Thank you for the informative reply, I appreciate it. As for DocTR being more efficient than EasyOCR that might be true, I forgot to mention that the approach I'm using is:
1. crop sections of PDF into small images.
2. Read with OCR.

This probably explains why I didn't have issues with EasyOCR. I forgot to crop the image when using DocTR 🤦‍♂️. I'll give it another shot with DocTR (especially now that you mentioned `detect_orientation=True` - thanks for that) but with cropped images this time.

I would love to share an example of the files that I'm working with, but I am unable to as they contain sensitive information about other people such as their name, SSN, and the university they study at. I work for a specific ministry/branch of my government that handles education and research in my country. These documents are sent in by students who study abroad, but still want the monthly government "allowance" so to speak. Students are paid to study here ranging from ~150€ to ~740€ depending on their situation (whether they live at home or on their own, how much their parents earn etc). I could perhaps censor the sensitive info or simply overwrite it with fake data. I think I'll try your suggested solutions first and if it still doesn't work I'll prepare some examples along with the code I ended up writing. :) 

You can see an example of an empty form [here](https://www.su.dk/media/1238/engelsk.pdf) (found on [su.dk](https://www.su.dk/udland/suiudlandetdokumentation/) at the bottom of the page section "`Studieaktivitet - krav til dokumentation`").

**Reply by felixT2K** [reply]
Hi again :)

Ok at the end if you know the information locations (coords) you need, you could use the whole `ocr_predictor` (det + reco) and export the results to a jsonlike dict with `json_output = result.export()`  (except handwritten is not supported by default yet).
Afterwards it should be easy to extract the required data.
Because under the hood we do the same : detection -> segmentations to boxes -> crop -> recognition

I understand this don't worry :)  
I only ever ask because an example makes it easier for us to understand what's going wrong.

**Reply by denivic** [reply]
Hey, so your suggestion worked. So far so good. I have a question in regards to some warnings though. Should I ignore these or is there a way to fix or remove them? 
![image](https://github.com/mindee/doctr/assets/32706842/5821e687-f75c-4fcb-9400-cfd7f29da9e5)

**Reply by felixT2K** [reply]
Hey :)

The first warning shouldn't appear anymore, looks like you're not up to date on the 'main' branch (version 0.7.0). 
(I know it's a pain currently but we have to wait that the model trainings finish ~november afterwards we will do the release) 😅 
About the second warning could you please verify that the boxes in the output sorted correct ?

**Reply by denivic** [reply]
It seems that it has not been updated on pip because upgrading like this:
`pip install --upgrade python-doctr`

Gives me messages about all the requirements being satisfied. I thought you implied that 0.7 was available and that the model itself will be updated in november, but I guess I misunderstood. 

As for whether the boxes are sorted, what do you mean by sorted correctly? Sorted by the detected word or? Is it supposed to be alphabetical?

**Reply by felixT2K** [reply]
@denivic Have you pulled the latest changes before with `git pull` ? :)

**Reply by denivic** [reply]
I just download via pip, but I'll just install manually. :)

EDIT: I mean pypi

**Reply by felixT2K** [reply]
About the sorting you can test it with:

```
result = predictor(doc)
full_text = result.render()
```
this will render the full text.
Than check that no words are reversed for example:

source: the quick brown fox.
output: the quick fox. brown

**Reply by denivic** [reply]
I installed the newest version successfully. Or so I thought. I'm using a virtual environment and made sure  that it was activated while trying to install the package, but it says the package doesn't exist. It also said the following:
```DEPRECATION: Loading egg at \\prod.sitad.dk\dfs\cu2411\i-drev\ufs-djos-ada\lis\lis - lukket\su udlandsrykker\sudocuchecker\lib\site-packages\python_doctr-0.7.0a0-py3.11.egg is deprecated.  pip 23.3 will enforce this behaviour change. A possible replacement is to use pip for package installation.```

Did I do something wrong? I ran this commando inside the venv (the folder was located on my desktop) `python setup.py install`. It's a little weird because it's listed when I do `pip list` inside the venv.

EDIT: Somehow I have 3 of them. 
![image](https://github.com/mindee/doctr/assets/32706842/19b4ba54-b251-4e06-9314-4426cd4b34b8)

**Reply by felixT2K** [reply]
The version is correct 👍🏼 
And no 0.7.0a0 is not deprecated 😅 
But of course it's weird that it's listed 3 times (it shouldn't)

**Reply by felixT2K** [reply]
```
cd doctr
pip3 install -e .[torch]
```

remove the other ones before :)

**Reply by denivic** [reply]
Hey @felixT2K sorry for the late reply.

I installed DocTR correctly now and checked the words are sorted correctly, which they seem to be. I'm still getting the `divide by zero` error however as well as these (probably due to Torchvision changing something - I believe it was in v0.13 but don't quote me on that 😄):
```
1. Using 'weights' as positional parameter(s) is deprecated since 0.13 and may be removed in the future. Please use keyword parameter(s) instead.

2. UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=None`.
  warnings.warn(msg)
```
As annoying as it may be, it is not that big of a deal honestly. There is probably some way to disable the warnings as well. 

What I would like to ask about instead is whether it is possible to have DocTR make the image straight if orientation != 0? Right now I'm using the code below, but would love to have a clean one-liner instead:
```py
def fix_orientation(self, image: np.ndarray, degree: int) -> np.ndarray:
    height, width = image.shape[:2]
    image_center = (width / 2, height / 2)

    rotation = cv2.getRotationMatrix2D(image_center, degree, 1)

    rad = math.radians(degree)
    sin = math.sin(rad)
    cos = math.cos(rad)
    b_w = int((height * abs(sin)) + (width * abs(cos)))
    b_h = int((height * abs(cos)) + (width * abs(sin)))

    rotation[0, 2] += ((b_w / 2) - image_center[0])
    rotation[1, 2] += ((b_h / 2) - image_center[1])

    return cv2.warpAffine(image, rotation, (b_w, b_h), flags=cv2.INTER_LINEAR)
```

**Reply by felixT2K** [reply]
Hi @denivic 👋🏼 ,

I found it where the torchvision weights warning is raised so i will open a PR to fix this soon. 👍🏼 

Currently we don't have any logic to fix the orientation in the input documents because this would need to modify the input which would need a bit refactoring of other components.
But feel free to open a feature request to add it :)

**Reply by felixT2K** [reply]
#1291  merged

**Reply by denivic** [reply]
@felixT2K I have encountered some problems with detecting rotating images. It worked sort of fine (although the analysis of **only** 4 images takes 71 seconds for some reason. Might be because I am testing on a laptop.

Anyway... Using the following code:
```py
def analyze(self, image_list: List[np.ndarray] | List[str] , output: Optional[str] = None) -> None | Dict[str, List[str]] | List[Dict[str, List[str]]]:
        if image_list is None:
            raise SUDocuExceptions.MissingInputException

        # Pre-trained model bruges.
        predictor = ocr_predictor(det_arch='db_resnet50_rotation',
                                  detect_orientation=True,
                                  pretrained=True,
                                  assume_straight_pages=False,
                                  export_as_straight_boxes=False,
                                  preserve_aspect_ratio=True)

        results = [predictor([img]) for img_list in image_list for img in img_list]

        if output is not None:
            for i, result in enumerate(results, start=1):
                with open(f'{output}\image_data_{i}.json', 'w', encoding='utf-8') as f:
                    f.write(dumps(result.export(), indent=4, ensure_ascii=False,  cls=NumpyEncoder))

            return
        else:
            return [result.export() for result in results] if isinstance(results, list) else results.export()
            
```
For using `ocr_predictor()` on a list of images loaded using `Document.from_images()`. This works fine although as mentioned is a little slow, but the issue I am facing currently is the fact that after analysis, I am double checking the orientation found by `ocr_predictor()` and for one image in particular it says the orientation is 0.0 when in reality it is not. This is the image:
![2](https://github.com/mindee/doctr/assets/32706842/c5a68260-4321-47da-9b3c-833ff966903c)

It is a bit of an outlandish example as no one in their right mind would send a document rotated like this, but it makes me worried that we might not catch all *legitimately* rotated images if it has issues with one that is so obviously rotated. Is there anything I can do to improve this?

Another thing, it says an image that has been rotated upside down has an orientation of -0.0. Shouldn't it have a rotation of 180 or -180 degrees and not -0.0? ~~Is this an intentional design choice to have it always display -0.0 when an image is upside down?~~ This does not seem to be the case because I manually flipped the image and now it says its orientation is 0.0. I thought the whole premise of this model was to detect rotation using the text contained within documents, but so far it isn't looking great. I mean it detects a decent amount correctly, but it also wrongly detects too many times. I might be doing something wrong so please correct me if I am.

Here is what it says for 4 images that are rotated (in order) 90, -180/180, 110 (around this value), -90, but `ocr_predictor` reports:
```py
# This one is correct I suppose, but isn't it supposed to be 90 if it's rotated 90 degrees to the right?
{'pages': [{'page_idx': 0, 'dimensions': (596, 842), 'orientation': {'value': -90.0, 'confidence': 1.0}}]}

# This one is incorrect. The image is flipped upside down. Should be -180 or 180 degress to left or right.
{'pages': [{'page_idx': 0, 'dimensions': (842, 531), 'orientation': {'value': -0.0, 'confidence': 1.0}}]}

# This one is also incorrect. This image is the one shown on GitHub and its orientation is definintely NOT 0.
{'pages': [{'page_idx': 0, 'dimensions': (926, 1025), 'orientation': {'value': 0, 'confidence': 1.0}}]}

# This one is correct.
{'pages': [{'page_idx': 0, 'dimensions': (596, 842), 'orientation': {'value': -90.0, 'confidence': 1.0}}]}
```

**Reply by felixT2K** [reply]
Hi @denivic 👋🏼,

You are correct in this case it should be 180 or -180 !
Maybe a bit confusing but the only differences between `db_resnet50_rotation` and `db_resnet50` is that it was trained on rotated documents (trough augmentation) - the next pytorch detection models we will release in 0.7.1 will all be trained on rotation augmented documents because it shows overall better results also for straight pages (more robust).

The orientation computation is done before the "image" is feed into the model:

https://github.com/mindee/doctr/blob/28e5375c399ae5dd146a1b7e13bd5831acb6dc2d/doctr/models/_utils.py#L29

This is a relative lazy approach and  as we can see requires some  "parameter tuning" specific to the document (for example if i change `ratio_threshold_for_lines` to < 3 it works for your provided example 😅 )

Normally we should grab the seg mask from the detection model build lines on the single segments and than compute the orientation angle. @charlesmindee wdyt ?

**Reply by denivic** [reply]
Ah I see. So in the upcoming version there won't be any need for tuning parameters and calling `estimate_orientation()` directly? Is that correctly understood?

I checked out the code by the way and I am a bit puzzled by this particular line:
https://github.com/mindee/doctr/blob/28e5375c399ae5dd146a1b7e13bd5831acb6dc2d/doctr/models/predictor/pytorch.py#L78

Is there any reason as to why the confidence has a hardcoded value of `1.0`? Isn't it a bit misleading as the point of a confidence value usually is to give you an idea of how accurate a given prediction is? To be fair I get that the `estimate_orientation()` method isn't exactly using an AI model to calculate the orientation so trying to calculate a "fake" confidence value so to speak is also kind of misleading. Was this the thought process behind this?

**Reply by felixT2K** [reply]
Hi @denivic 👋🏼,

No, you have misunderstood me a bit.
The models that were trained with rotations are able to detect rotated text better, but not the angle of the rotation.
For this only the mentioned function is responsible, which as you can see is not very stable.

My idea would be to determine the rotation angle of the whole document using the segmentation mask, which we get from the detection model.

On the other hand if `straighten_pages=True` is passed to the `ocr_predictor` we would need to forward 2 times trough the model 🤔 

In the second case, I agree with you, I'll have to see what the intention was behind it.

**Reply by denivic** [reply]
Hey @felixT2K 

Thank you for your continued support, it is much appreciated. If you don't mind, I would love to hear a bit more about how to use a segmentation mask. How would this look in code? If this is already explained somewhere in the documentation (I wasn't able to find it though) then do you mind linking me to it? :) 

Kind regards,
Denis

**Reply by denivic** [reply]
Also is there any ETA on the new update, which includes the improved model? There has recently been a development in my project such that the deadline has become a bit tighter, if you know what I mean.

**Reply by felixT2K** [reply]
Hi @denivic i am on it see https://github.com/mindee/doctr/pull/1336
I will update you soon :)

**Reply by felixdittrich92** [reply]
@denivic It's still in progress but already testable so you can checkout to the branch and test it i'm happy about every feedback :)

**Reply by felixdittrich92** [reply]
Btw. for the 0.7.1 release we don't have a fixed deadline but the goal is to do the release end of november/ start december

**Reply by denivic** [reply]
> Hi @denivic i am on it see #1336 I will update you soon :)

Amazing, thank you so much! 😁

> Btw. for the 0.7.1 release we don't have a fixed deadline but the goal is to do the release end of november/ start december

Cool, I'm looking forward to it.

**Reply by felixdittrich92** [reply]
@denivic Could you test it on the branch ? 

It should work much better for rotated samples in a range of (-90 - 90 degrees) 

- will further improve with the new models (because now depends on the detection results)
- upside down detection is still a todo task

**Comment by denivic** [comment]
@felixdittrich92 I have tested #1336 on a sample size of 21 documents with random angles between `-90` and `90` and it seems to not be working as expected. I might be doing something wrong, but this is what I'm getting:
```py
1. Orientation found: 14.845458984375, Actual orientation: -3
2. Orientation found: 10.808235168457031, Actual orientation: 3
3. Orientation found: -73.0724868774414, Actual orientation: 19
4. Orientation found: 39.805572509765625, Actual orientation: -70
5. Orientation found: 86.9471173286438, Actual orientation: 48
6. Orientation found: 12.90740966796875, Actual orientation: 71
7. Orientation found: -42.41420364379883, Actual orientation: -32
8. Orientation found: 0.6437454223632812, Actual orientation: -28
9. Orientation found: -6.009005546569824, Actual orientation: 15
10. Orientation found: -49.39870071411133, Actual orientation: -4
11. Orientation found: -52.177249908447266, Actual orientation: 4
12. Orientation found: 48.90570068359375, Actual orientation: 6
13. Orientation found: -45.364933013916016, Actual orientation: 74
14. Orientation found: 26.113914489746094, Actual orientation: -78
15. Orientation found: 45.88140106201172, Actual orientation: 85
16. Orientation found: 79.5085220336914, Actual orientation: 83
17. Orientation found: 60.80251693725586, Actual orientation: -7
18. Orientation found: 0, Actual orientation: -40
19. Orientation found: -8.310276985168457, Actual orientation: 62
20. Orientation found: -14.444034576416016, Actual orientation: 90
21. Orientation found: -9.090276718139648, Actual orientation: -43
```
[Using these methods](https://gist.github.com/denivic/5ceb5fa48cd5370d341e16945563f5e9). It could possibly be that I'm iterating through my list of randomly selected angles the wrong way with respect to the list of images, but I don't think that I am though unless `ocr_predictor` returns in an unordered fashion then these results might be wrong. I can see that some of these match (or are at the very least *very* close) such as the actual rotation of number 5 matching the orientation found for number 12. I might be reading too much into this though, it could very well be a coincidence.

EDIT: Well... Seems like some of the images are rotated more than 90 or -90 degrees for some reason causing them to be almost upside down, which is weird because I checked the images manually before loading them using `DocumentFile.from_images()`. Some of these results are still accurate (accurate in the sense that the images are actually within the -90 and 90 range yet the wrong orientation is reported). I'll re-do these tests tomorrow and verify manually that each and every image that is being tested is within -90 and 90.

**Reply by felixdittrich92** [reply]
Hi @denivic :wave: 

I have tested it also again:
```
orientation: -33.690067291259766, rnd: -33
orientation: -20.95577621459961, rnd: -21
orientation: 60.99358367919922, rnd: 61
orientation: -60.64224624633789, rnd: -61
orientation: 58.920963287353516, rnd: 59
orientation: -86.94129943847656, rnd: -87
orientation: -45.0, rnd: -45
orientation: -12.05399227142334, rnd: -12
orientation: 0, rnd: 2
orientation: 17.10272979736328, rnd: 17
orientation: 37.14668655395508, rnd: 37
orientation: 30.004924774169922, rnd: 30
orientation: -18.060474395751953, rnd: -18
orientation: 58.96489143371582, rnd: 59
orientation: 32.927860260009766, rnd: 33
orientation: -72.13231658935547, rnd: -72
orientation: 13.080741882324219, rnd: 13
orientation: -88.05191802978516, rnd: -88
orientation: -14.123187065124512, rnd: -14
orientation: 7.8533172607421875, rnd: 7
```
I will post my test code in a few minutes :)


```

import time

from doctr.io import DocumentFile
from doctr.models import ocr_predictor
from doctr.utils.geometry import rotate_image
import random
import requests


bytes_data = requests.get("https://www.allianzdirect.de/dam/documents/home/Versicherungsbedingungen-08-2021.pdf").content

start = time.time()
doc = DocumentFile.from_pdf(bytes_data)
rnds = [random.randint(-90, 90) for _ in range(len(doc))]
doc = [rotate_image(page, rnd, expand=True) for page, rnd in zip(doc, rnds)]

model = ocr_predictor(
    det_arch="db_resnet50_rotation",
    reco_arch="parseq",
    pretrained=True,
    det_bs=8,
    reco_bs=1024,
    assume_straight_pages=False,
    straighten_pages=True,
    detect_orientation=True,
)#.cuda().half()  # uncomment if you have a GPU

result = model(doc)
print(time.time() - start)
result.show()
json_res = result.export()

for res, rnd in zip(json_res['pages'], rnds):
    orientation = res['orientation']['value']
    print(f"orientation: {orientation}, rnd: {rnd}")
```

**Reply by denivic** [reply]
Damn, I knew I fucked up somehow. 😂

I am looking forward to checking your code out. Thank you!

**Reply by felixdittrich92** [reply]
The detection results from `straighten_pages=True` are a bit broken after the changes.
The new models should fix this and bring also further accuracy on orientation detection.

@denivic code snippet added

EDIT: upside down is still an open issue (We would need to train an classifier for this that we can afterwards correct the angle correctly - but that would be more a ticket for 0.8.0)

**Reply by denivic** [reply]
@felixdittrich92 I have downloaded the most recent version of DocTR [with the changes you made](https://github.com/felixdittrich92/doctr/tree/orientation-seq-prototype) and tried to run [the following code](https://gist.github.com/denivic/40e2f44ab50e816580d050c8522e440c). However if I run it using `Document.from_images()` (that's all `multiload` is - a wrapper method basically) I get:
```py
orientation: -90, rnd: 32
orientation: 0, rnd: 2
orientation: 0, rnd: -36
orientation: 0, rnd: -2
```
I thought it might have something to do with images vs pdfs, so I tried to run it using just a single PDF but for some reason I get the following error:
```py
  File "c:\Users\<user>\Desktop\Projekter\SU Udlandsrykker\SUDocuChecker\Lib\site-packages\python_doctr-0.7.1a0-py3.11.egg\doctr\io\pdf.py", line 41, 
in <listcomp>
AttributeError: 'PdfPage' object has no attribute 'render'. Did you mean: 'render_to'?
```
I wanted to try and change `render` to `render_to` myself, but I have no idea how .egg files work and what they do. Any ideas as to what might be going wrong?

**Reply by felixT2K** [reply]
Hi @denivic 👋🏼,

About `render` it looks like you have an outdated depedency of `pypdfium2` installed you should update it :)

Anything is weird because i have tested it on TF and PT with different images and pdfs and everything was mostly correct.

Which detection model do you use ?

**Reply by felixT2K** [reply]
Can you please check that you get the same results with the script i have provided above ? (Will differ a bit because of the random rotation but the orientations should be mostly correct)

**Reply by denivic** [reply]
Oh, right. Updated the package now and tested. It works!

Also, I forgot to mention, but the first thing I did was try and run your script and it worked perfectly. Now I just have to test and see if it works for my images as well.

**Reply by felixT2K** [reply]
Nice 👍🏼 

Let me know if it works happy about every feedback :)
And keep in mind the full potential it will only show its full potential once we have trained the detection models. (runs will start soon)

**Reply by denivic** [reply]
@felixT2K Ok so I have made sure all dependencies are updated and tried running the exact same code as shown in the link above but instead of loading a pdf using `from_pdfs`, I loaded the same PDF, which has been converted to an image, using `from_image`. I get weird results. First, it always gives me a value of 0 despite it working with a pdf:
```py
orientation: 0, rnd: 54
```
And secondly, no matter what image I use I get "divide by zero error encountered". You mentioned something about the order of words previously so I checked out the results by exporting the `json` data to a textfile and it seems to all be in order. I tried using a random image that I downloaded from the internet. While it didn't give me a `divide by zero error` like previously, it still gave me wrong results:
```py
orientation: 0, rnd: 75
```
Any ideas?

**Reply by denivic** [reply]
This is the image I used from the internet by the way (if you wanna test yourself):
![dog](https://github.com/mindee/doctr/assets/32706842/a51a9c20-ab47-4981-ab66-58a2748b0e22)

**Reply by felixT2K** [reply]
Ok that any random images without textual content doesn't works should be clear ^^
Would it be possible to provide any pdf which doesn't worked for you while testing ? 
In this case i can try to reconstruct it.

**Reply by denivic** [reply]
I get that it's an OCR predictor, but I suppose I assumed (incorrectly I guess) that it'd still compute the orientation despite no text being present and just output `None` in regards to text. Is it computing the orientation using the text itself? 

Here is the PDF that doesn't work when I convert it to a `.png` image:  https://www.su.dk/media/1239/dansk.pdf

**Reply by felixdittrich92** [reply]
Correct we compute the orientation on the median line of the segmentation map which we get from the detection module

**Reply by felixdittrich92** [reply]
I have tested it with your provided pdf and duplicated it 20 times to apply random rotation.

Code PyTorch:

```
import random
import time

import requests

from doctr.io import DocumentFile
from doctr.models import ocr_predictor
from doctr.utils.geometry import rotate_image

#bytes_data = requests.get(
#    "https://www.allianzdirect.de/dam/documents/home/Versicherungsbedingungen-08-2021.pdf"
#).content

start = time.time()
# doc = DocumentFile.from_pdf("/home/felix/Desktop/vitstr_paper.pdf")
#doc = DocumentFile.from_pdf(bytes_data)
# doc = DocumentFile.from_images(['/home/felix/Desktop/5.png']) # * 20)
doc = DocumentFile.from_pdf('/home/felix/Desktop/dansk.pdf') * 20
rnds = [random.randint(-90, 90) for _ in range(len(doc))]
# rnds = [random.choice([90, -90]) for _ in range(len(doc))]
doc = [rotate_image(page, rnd, expand=True) for page, rnd in zip(doc, rnds)]

model = ocr_predictor(
    det_arch="db_resnet50_rotation",
    reco_arch="parseq",
    pretrained=True,
    det_bs=8,
    reco_bs=1024,
    assume_straight_pages=False,
    straighten_pages=True,
    detect_orientation=True,
).cuda().half()  # uncomment if you have a GPU

result = model(doc)
print(time.time() - start)
result.show()
json_res = result.export()

for res, rnd in zip(json_res["pages"], rnds):
    orientation = res["orientation"]["value"]
    print(f"orientation: {orientation}, rnd: {rnd}")
```

Results:
```
orientation: -63, rnd: -63
orientation: -21, rnd: -21
orientation: -17, rnd: -17
orientation: -55, rnd: -55
orientation: 85, rnd: 85
orientation: -14, rnd: -14
orientation: -47, rnd: -47
orientation: -83, rnd: -83
orientation: -54, rnd: -54
orientation: 20, rnd: 20
orientation: -82, rnd: -82
orientation: 33, rnd: 33
orientation: -4, rnd: -4
orientation: 5, rnd: 5
orientation: 51, rnd: 51
orientation: 61, rnd: 61
orientation: 34, rnd: 34
orientation: 31, rnd: 31
orientation: -4, rnd: -4
orientation: -22, rnd: -22
```

Code Tensorflow:
```
import random
import time

import requests

from doctr.io import DocumentFile
from doctr.models import ocr_predictor
from doctr.utils.geometry import rotate_image

#bytes_data = requests.get(
#    "https://www.allianzdirect.de/dam/documents/home/Versicherungsbedingungen-08-2021.pdf"
#).content

start = time.time()
# doc = DocumentFile.from_pdf("/home/felix/Desktop/vitstr_paper.pdf")
#doc = DocumentFile.from_pdf(bytes_data)
# doc = DocumentFile.from_images(['/home/felix/Desktop/5.png']) # * 20)
doc = DocumentFile.from_pdf('/home/felix/Desktop/dansk.pdf') * 20
rnds = [random.randint(-90, 90) for _ in range(len(doc))]
# rnds = [random.choice([90, -90]) for _ in range(len(doc))]
doc = [rotate_image(page, rnd, expand=True) for page, rnd in zip(doc, rnds)]

model = ocr_predictor(
    det_arch="linknet_resnet18_rotation",
    reco_arch="parseq",
    pretrained=True,
    det_bs=8,
    reco_bs=1024,
    assume_straight_pages=False,
    straighten_pages=True,
    detect_orientation=True,
)#.cuda().half()  # uncomment if you have a GPU

result = model(doc)
print(time.time() - start)
result.show()
json_res = result.export()

for res, rnd in zip(json_res["pages"], rnds):
    orientation = res["orientation"]["value"]
    print(f"orientation: {orientation}, rnd: {rnd}")
```

Results:
```
orientation: 38, rnd: 38
orientation: -59, rnd: -59
orientation: 19, rnd: 19
orientation: -24, rnd: -24
orientation: -90, rnd: 90
orientation: -16, rnd: -16
orientation: -4, rnd: -4
orientation: -11, rnd: -11
orientation: 48, rnd: 48
orientation: -21, rnd: -21
orientation: -34, rnd: -34
orientation: 39, rnd: 39
orientation: -36, rnd: -36
orientation: 57, rnd: 57
orientation: -25, rnd: -25
orientation: 2, rnd: 2
orientation: 52, rnd: 52
orientation: 65, rnd: 65
orientation: 40, rnd: 40
orientation: 39, rnd: 39
```

Everything looks as expected on my side :sweat_smile: 

Are you sure that you are checked out correctly into my changes ?

```
(doctr-dev) felix@felix-GS66-Stealth-11UH:~/Desktop/doctr$ git branch
  benchmark-docs
  main
* orientation-seq-prototype
  tf-detection-augmentations
```

* marked the branch you need to checkout

**Reply by felixdittrich92** [reply]
Otherwise, please be patient. I think @odulcy-mindee  can review the rest on Friday, so we can merge it next week at the latest.
Then you could pull the changes directly from main branch.

**Reply by denivic** [reply]
I may not have the correct version then. I apologize for the confusion and waste of time. I'll have to double check tomorrow and try to verify using the code above. I'll do a 100% clean re-install of DocTR and its dependencies. I'll let you know. Again, sorry. :)

**Comment by felixdittrich92** [comment]
> I may not have the correct version then. I apologize for the confusion and waste of time. I'll have to double check tomorrow and try to verify using the code above. I'll do a 100% clean re-install of DocTR and its dependencies. I'll let you know. Again, sorry. :)

It's all ok :)

Ensure that you are on the correct branch : `orientation-seq-prototype`

**Reply by felixdittrich92** [reply]
You can simply check it by : `git branch`

**Reply by denivic** [reply]
@felixdittrich92 Alright so it turns out I _did_ have the latest version installed, but something somewhere went wrong. I uninstalled and then reinstalled DocTR and all its dependencies to test once again and it works without issue now (besides small inconcistencies, but it's waaaaay better than my previous results still).
```py
orientation: 41, rnd: -48
orientation: 53, rnd: 53
orientation: 69, rnd: 69
orientation: 78, rnd: 78
orientation: 41, rnd: 41
orientation: 41, rnd: 41
orientation: 54, rnd: 54
orientation: 75, rnd: 75
orientation: 77, rnd: 77
orientation: 63, rnd: 63
```
I noticed I had duplicate installations of both DocTR (same version though) and its dependencies (also same version). Does `pip` not detect that there's a package with the same version already installed? 

Anyway, I'm looking forward to the next release. For now this is great though. Thank you so much for the help. :)

**Reply by felixT2K** [reply]
Normally it should 😅 
Nice that it now works on your side also 👍🏼 
The results will further improve with the new detection models :)

**Reply by denivic** [reply]
Hello @felixdittrich92 

I've been watching #1336 closely and I'm happy that you've put so much work in. Appreciate it.

I was just wondering if there's an ETA on when 0.8 will release? 😄

**Reply by felixT2K** [reply]
Hey @denivic 👋🏼 

Thanks for your kind words 😄 

I depends a bit how fast we get the following ticket / runs done #1364 
So personally i hope that we are done with it in january afterwards only a bit minor stuff left which should be not take more then 1 week and we can do the release

**Reply by felixT2K** [reply]
merged :)

**Comment by nehashinde-canopy** [comment]
I am getting an error while running docTR:
`predictor = detection_predictor('db_resnet50_rotation', pretrained=True, assume_straight_pages=False, preserve_aspect_ratio=True)`
error:
ValueError: unknown architecture 'db_resnet50_rotation'
I have installed pytorch version of doctr using pip install "python-doctr[torch]"
Kindly help.

**Reply by felixdittrich92** [reply]
`db_resnet50_rotation` is removed in v0.8.0 it is now `db_resnet50`

**Comment by nehashinde-canopy** [comment]
Is there no way I can detect the rotation angle in the image?

On Fri, Mar 1, 2024 at 3:38 PM Felix Dittrich ***@***.***>
wrote:

> db_resnet50_rotation is removed in v0.8.0 it is now db_resnet50
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mindee/doctr/discussions/1283#discussioncomment-8641216>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BGKCWN36SIBUG4QJR3NHQKTYWBHQ3AVCNFSM6AAAAAA3XEETMCVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DMNBRGIYTM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

-- 
This message is intended for the use of the addressee and may contain 
information that is privileged and/or confidential. If you are not the 
intended recipient, you are hereby notified that any dissemination, 
distribution or copying of the information contained in this message is 
strictly unauthorized and prohibited. If you have received this message in 
error, please notify the sender by reply e-mail and delete the message from 
your system.

**Reply by felixdittrich92** [reply]
That's possible with the `ocr_predictor` instance and `detect_orientation=True` :)

**Comment by nehashinde-canopy** [comment]
Thanks I'll try this. Is there a way I can get a key value kie extraction
like:
{
Name: Neha,
Dob: 26-03
}
I get only "words" as a class when I implement KIE. I need keys and values
from the document.

Thanks,
Neha

On Fri, 1 Mar, 2024, 7:25 pm Felix Dittrich, ***@***.***>
wrote:

> That's possible with the ocr_predictor instance and
> detect_orientation=True :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mindee/doctr/discussions/1283#discussioncomment-8643441>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BGKCWNZUZNZTSTF66SF64C3YWCCGVAVCNFSM6AAAAAA3XEETMCVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DMNBTGQ2DC>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

-- 
This message is intended for the use of the addressee and may contain 
information that is privileged and/or confidential. If you are not the 
intended recipient, you are hereby notified that any dissemination, 
distribution or copying of the information contained in this message is 
strictly unauthorized and prohibited. If you have received this message in 
error, please notify the sender by reply e-mail and delete the message from 
your system.
